### ISSUE TITLE
Add ability to build solve sequence stack in the Executioner

### ISSUE THREAD

**Issue by lindsayad** [idaholab/moose#25416]
## Reason
This will make things like #25327 easy (hopefully). 

## Design
These were the initially suggested APIs in https://github.com/idaholab/moose/pull/25328#discussion_r1315145080:

```c++
const std::vector<std::pair<ExecFlagType, IterationInfo>> & solveStack() const;
std::vector<std::pair<ExecFlagType, IterationInfo>> & solveStack();
virtual const IterationInfo & getIterationNumber(const ExecFlagType & iteration_type) const;
```

but we will likely forego the `ExecFlagType` part of that API for something else. The stack of iteration information will be created in a way like @dschwen proposed:

```
  {
    const SolveLoop solve_loop(*this, this->_counter);
    <solve loop code>
  }
```
and
```c++
class SolveLoop
{
public:
  SolveLoop(Executioner & ex, const unsigned int & counter) : _ex(ex)
    {
      _ex.solveStack().push_back(&counter);
    }

  ~SolveLoop() { _ex.solveStack().pop_back(); }

private:
  Executioner & _ex;
};
```

## Impact
Hopefully a convenient way to codify what solve point we are at in a simulation, and consequently to delineate output when that delineation is desired.

**Comment by lindsayad** [comment]
Product of discussion with @recuero @lynnmunday @dschwen and @GiudGiud. We can and will of course iterate on this

**Comment by YaqiWang** [comment]
Are you going to abandon Executor system?

**Comment by GiudGiud** [comment]
No executors will create exec flags and will be able to create loops too. The same problem of know when an output point is exists for executors

**Comment by lindsayad** [comment]
Yes many objects will likely create solve loops. There just needs to be a central place where knowledge of the stack of solve loops is. And there is only one executioner so it’s logical choice to me
